src/box/vy_scheduler.c

```diff
@@ -1996,47 +2019,45 @@ vy_task_complete(struct vy_task *task)
 	return -1;
 }
 
+static void
+vy_scheduler_finish_tasks(struct vy_scheduler *scheduler, int *tasks_done,
+			  int *tasks_failed)
+{
+	*tasks_done = 0;
+	*tasks_failed = 0;
+	while (!stailq_empty(&scheduler->processed_tasks)) {
+		struct stailq tasks = STAILQ_INITIALIZER(tasks);
+		struct vy_task *task, *next;
+		stailq_concat(&tasks, &scheduler->processed_tasks);
+		stailq_foreach_entry_safe(task, next, &tasks, in_processed) {
+			if (vy_task_complete(task) == 0)
+				(*tasks_done)++;
+			else
+				(*tasks_failed)++;
+			vy_worker_pool_put(task->worker);
+			vy_task_delete(task);
+		}
+	}
+}
+
 static int
 vy_scheduler_f(va_list va)
 {
 	struct vy_scheduler *scheduler = va_arg(va, struct vy_scheduler *);
 
-	while (scheduler->scheduler_fiber != NULL) {
-		struct stailq processed_tasks;
-		struct vy_task *task, *next;
-		int tasks_failed = 0, tasks_done = 0;
+	while (!fiber_is_cancelled()) {
+		struct vy_task *task;
+		int tasks_done, tasks_failed;
 
 		fiber_check_gc();
-		/* Get the list of processed tasks. */
-		stailq_create(&processed_tasks);
-		stailq_concat(&processed_tasks, &scheduler->processed_tasks);
-
-		/* Complete and delete all processed tasks. */
-		stailq_foreach_entry_safe(task, next, &processed_tasks,
-					  in_processed) {
-			if (vy_task_complete(task) != 0)
-				tasks_failed++;
-			else
-				tasks_done++;
-			vy_worker_pool_put(task->worker);
-			vy_task_delete(task);
-		}
+		vy_scheduler_finish_tasks(scheduler, &tasks_done,
+					  &tasks_failed);
 		/*
 		 * Reset the timeout if we managed to successfully
 		 * complete at least one task.
 		 */
-		if (tasks_done > 0) {
+		if (tasks_done > 0)
 			scheduler->timeout = 0;
-			/*
-			 * Task completion callback may yield, which
-			 * opens a time window for a worker to submit
-			 * a processed task and wake up the scheduler
-			 * (via scheduler_async). Hence we should go
-			 * and recheck the processed_tasks in order not
-			 * to lose a wakeup event and hang for good.
-			 */
```

Comment:
Please don't delete the comment. Move it to to `vy_scheduler_complete_tasks`.

Reply:
Reword it a bit for the new context:

```diff
diff --git a/src/box/vy_scheduler.c b/src/box/vy_scheduler.c
index 01da4efca..2874c4bb1 100644
--- a/src/box/vy_scheduler.c
+++ b/src/box/vy_scheduler.c
@@ -2025,6 +2025,11 @@ vy_scheduler_complete_tasks(struct vy_scheduler *scheduler, in>
 {
        *tasks_done = 0;
        *tasks_failed = 0;
+       /*
+        * Task completion callback may yield thus we have the loop to
+        * completely drain processed tasks queue. So that callers may
+        * wait for new tasks in the queue after calling this function.
+        */
        while (!stailq_empty(&scheduler->processed_tasks)) {
                struct stailq tasks = STAILQ_INITIALIZER(tasks);
                struct vy_task *task, *next;
```